### PR TITLE
feat: support batched graphql operations

### LINKED ISSUE #510: Support batched GraphQL queries
## Environment

| Name    | Version |
| ------- | ------- |
| msw     | 0.24.2   |
| browser | chrome Version 87.0.4280.88 (Official Build) (x86_64) |
| OS      | macOS Big Sur 11.0.1 (20B29)  |

## Request handlers

```js

export const handlers = [
  graphql.operation((req, res, ctx) => {
    console.log({ req, res, ctx });
    return res(
      ctx.data({
        user: {
          username: 'test',
          firstName: 'John',
        },
      })
    );
  }),

  graphql.query('getTest', (req, res, ctx) => {
    return res(
      ctx.data({
        user: {
          username: 'test',
          firstName: 'John',
        },
      })
    );
  }),
];

```

## Actual request

https://www.apollographql.com/docs/link/links/batch-http/
Using Apollo Client to perform a normal batch query

ie.
```
import { BatchHttpLink } from "apollo-link-batch-http";
const link = new BatchHttpLink({ uri: "/graphql" });
```

## Current behavior

batch queries are not getting intercepted but if I switch from `BatchHttpLink` to a regular `HttpLink`, the interception works as expected.

## Expected behavior

batch queries get intercepted

### PR DESCRIPTION
**Pull request by msutkowski**
Closes #510

### CODE REVIEW

**Comment by msutkowski** [comment]
This actually needs more work. I finally set up an environment to actually test the batch link behavior and ran into issues. Will update later on :)

**Comment by dkhuntrods** [comment]
@msutkowski 👋 

I've [done some work on this](https://github.com/BulbEnergy/msw/pull/1) (branched from an earlier version as we need to support older TS versions) as we needed the functionality ASAP.

Hopefully it's of help - we'd much rather to be able to use and support original package!

/cc @kettanaito

**Comment by kettanaito** [comment]
Thank you for providing that implementation reference, @dkhuntrods! 

With the recent changes to the request handlers, this pull request needs a little adjustment. I've tried to structure the steps that a request handler performs, so the new logic we add will fall under one of the steps (most likely a parsing phase). I'd be happy to give a hand in aligning these changes with the latest state of the repository. 

A huge thank you, @msutkowski, for preparing this. Let me know if you need help with rebasing this branch.

**Comment by kettanaito** [comment]
Hey, @msutkowski. During our discussion, I've mentioned a utility that can differentiate between REST and GraphQL requests. Here it is:

https://github.com/mswjs/msw/blob/655c93f83b7fcebe9a9093cacd7d891939028d29/src/utils/internal/parseGraphQLRequest.ts#L139-L172

We are already using it in the GraphQL request handler during the parsing phase:

https://github.com/mswjs/msw/blob/655c93f83b7fcebe9a9093cacd7d891939028d29/src/handlers/GraphQLHandler.ts#L107-L109

That parsing result is later taken into account in the `predicate` step that decides whether a request should be treated as a GraphQL request and be handled by the respective handler:

https://github.com/mswjs/msw/blob/655c93f83b7fcebe9a9093cacd7d891939028d29/src/handlers/GraphQLHandler.ts#L121-L124

It seems that no extra logic should be necessary as a part of the batched GraphQL queries support. As always, I may be wriong. Would appreciate your expertise on this. Thank you.

File: src/utils/getResponse.ts [hunk 1]
```diff
@@ -56,8 +56,15 @@ export const getResponse = async <
       handler: null,
       response: null,
     }
+  } else if (relevantHandlers.length > 1) {
+    // TODO: this is probably not the best approach... we need some way to limit the result set of relevantHandlers to 1 of each type.
+    // Is this really even possible though? Will one request ever _really_ match multiple handlers?
```
Comment (kettanaito) [inline src/utils/getResponse.ts:61]: If a request has a matching handler and that handler returns a mocked response, _only that handler_ will be used. Even if there are other handlers that match, they are resolved in the order they were attached. A single request cannot respect multiple matching handlers, this is a no-op. 

Why do we need to have this logic, though?

File: test/graphql-api/utils/executeOperation.ts [hunk 2]
```diff
@@ -83,3 +66,59 @@ export const executeOperation = async (
...
+) => {
+  const { uri = HOSTNAME, method = 'POST' } = options || {}
+  const url = new URL(uri)
+  const urlString = url.toString()
+
+  const results = await Promise.all(
+    payload.map(({ query, variables }) =>
+      performRequest({
+        page,
+        urlString,
```
Comment (kettanaito) [inline test/graphql-api/utils/executeOperation.ts:116]: As these requests will be to the same URL, `performRequest` will resolve on the first match. This results in both results having the same JSON body.

**Comment by kettanaito** [comment]
@msutkowski, hey 👋 Quite a bunch of things has changed around request handlers since this improvement was opened. I think you are right to consider opening a new pull request, it'll be less hassle than trying to migrate the current state. 

Thank you so much for working on this one!

**Comment by otaciliolacerda** [comment]
Hi all, by the state of this branch can I assume `msw` currently does not support batch requests?

**Comment by kettanaito** [comment]
## Released: v2.1.3 🎉

This has been released in v2.1.3!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.1.3)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.1.3)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Pull request closed without merging**